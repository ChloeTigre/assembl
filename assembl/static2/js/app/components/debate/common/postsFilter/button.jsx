// @flow

import * as React from 'react';
import { Button, MenuItem } from 'react-bootstrap';
import { Translate } from 'react-redux-i18n';

type PostsFilterButtonProps = {
  i18nTitle: string,
  onClick: () => void
};

class DumbPostsFilterButton extends React.Component<PostsFilterButtonProps> {
  handleClick = (event: any) => {
    event.preventDefault();
    event.stopPropagation();
  };

  render() {
    const { onClick, i18nTitle, ...buttonProps } = this.props;
    return (
      <MenuItem onClick={this.handleClick}>
        <Button onClick={onClick} {...buttonProps}>
          <Translate value={i18nTitle} />
        </Button>
      </MenuItem>
    );
  }
}

export default DumbPostsFilterButton;