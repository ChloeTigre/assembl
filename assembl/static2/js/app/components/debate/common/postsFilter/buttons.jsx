// @flow
import * as React from 'react';
import { MenuItem } from 'react-bootstrap';

type Props = {
  children: React.Node
};

class DumbPostsFilterButtons extends React.Component<Props> {
  handleClick = (event: any) => {
    event.preventDefault();
    event.stopPropagation();
  };

  render() {
    const { children } = this.props;
    return <MenuItem onClick={this.handleClick}>{children}</MenuItem>;
  }
}

export default DumbPostsFilterButtons;