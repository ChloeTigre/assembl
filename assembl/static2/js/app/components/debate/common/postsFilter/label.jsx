// @flow
import * as React from 'react';
import { MenuItem } from 'react-bootstrap';
import { Translate } from 'react-redux-i18n';

function noop(event) {
  event.preventDefault();
  event.stopPropagation();
}

type Props = {
  labelMsgId: string
};

class DumbPostsLabelMenuItem extends React.Component<Props> {
  render() {
    return (
      <MenuItem onClick={noop} disabled>
        <strong>
          <Translate value={this.props.labelMsgId} />
        </strong>
      </MenuItem>
    );
  }
}

export default DumbPostsLabelMenuItem;