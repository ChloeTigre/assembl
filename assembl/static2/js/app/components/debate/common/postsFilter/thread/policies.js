// @flow
import {
  chronologicalLastPolicy,
  fullDisplayPolicy,
  myPostsAndAnswersFilterPolicy,
  onlyMyPostFilterPolicy,
  popularityPolicy,
  reverseChronologicalFlatPolicy,
  reverseChronologicalLastPolicy,
  reverseChronologicalTopPolicy,
  summaryDisplayPolicy
} from '../policies';

export const postsOrderPolicies: PostsOrderPolicy[] = [
  reverseChronologicalTopPolicy,
  reverseChronologicalLastPolicy,
  chronologicalLastPolicy,
  reverseChronologicalFlatPolicy,
  popularityPolicy
];

export const postsDisplayPolicies: PostsDisplayPolicy[] = [fullDisplayPolicy, summaryDisplayPolicy];

export const postsFiltersPolicies: PostsFilterPolicy[] = [onlyMyPostFilterPolicy, myPostsAndAnswersFilterPolicy];