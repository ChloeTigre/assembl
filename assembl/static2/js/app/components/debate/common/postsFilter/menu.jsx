// @flow
import * as React from 'react';
import { DropdownButton } from 'react-bootstrap';
import debounce from 'lodash/debounce';
import classNames from 'classnames';
import { I18n } from 'react-redux-i18n';
import { connect } from 'react-redux';

import PostsFilterMenuItem from './menuItem';
import PostsFilterButton from './button';
import PostsFilterButtons from './buttons';
import PostsFilterLabelMenuItem from './label';
import { withScreenHeight } from '../../../common/screenDimensions';
import { setThreadPostsFilterPolicies } from '../../../../actions/threadFilterActions';
import { defaultDisplayPolicy, defaultOrderPolicy, postsDisplayPolicies, postsOrderPolicies } from './policies';

type Props = {
  postsDisplayPolicy: PostsDisplayPolicy,
  postsOrderPolicy: PostsOrderPolicy,
  screenHeight: number,
  setPostsFilterPolicies: (postsDisplay: PostsDisplayPolicy, postsOrder: PostsOrderPolicy) => void
};

type State = {
  selectedPostsDisplayPolicy: PostsDisplayPolicy,
  selectedPostsOrderPolicy: PostsOrderPolicy,
  sticky: boolean
};

export class DumbPostsFilterMenu extends React.Component<Props, State> {
  componentWillMount(): void {
    const { postsDisplayPolicy, postsOrderPolicy } = this.props;
    this.setState({
      sticky: true,
      selectedPostsDisplayPolicy: postsDisplayPolicy,
      selectedPostsOrderPolicy: postsOrderPolicy
    });
  }

  componentDidMount() {
    window.addEventListener('scroll', this.setButtonPosition);
  }

  componentWillUnmount = () => {
    window.removeEventListener('scroll', this.setButtonPosition);
  };

  setButtonPosition = debounce(() => {
    const { screenHeight } = this.props;
    if (window.pageYOffset > screenHeight - 60) {
      // Show the button when we scrolled minimum the height of the window.
      this.setState({ sticky: true });
    } else {
      this.setState({ sticky: false });
    }
  }, 100);

  selectPostsDisplayPolicy = (postsDisplayPolicy: any) => {
    // FIXME PostsDisplayPolicy
    this.setState({ selectedPostsDisplayPolicy: postsDisplayPolicy });
  };

  selectPostsOrderPolicy = (postsOrderPolicy: any) => {
    // FIXME PostsOrderPolicy
    this.setState({ selectedPostsOrderPolicy: postsOrderPolicy });
  };

  resetPolicies = () => {
    this.selectPostsDisplayPolicy(defaultDisplayPolicy);
    this.selectPostsOrderPolicy(defaultOrderPolicy);
    this.props.setPostsFilterPolicies(defaultDisplayPolicy, defaultOrderPolicy);
  };

  savePolicies = () => {
    this.props.setPostsFilterPolicies(this.state.selectedPostsDisplayPolicy, this.state.selectedPostsOrderPolicy);
  };

  render() {
    const { selectedPostsDisplayPolicy, selectedPostsOrderPolicy } = this.state;
    const { sticky } = this.state;
    return (
      <div className={classNames(['posts-filter-button', sticky ? 'sticky' : null])}>
        <DropdownButton
          pullRight
          title={
            <img
              height={24}
              width={24}
              src="/static2/img/icons/black/filter.svg"
              alt={I18n.t('debate.thread.sortFilterPosts')}
              title={I18n.t('debate.thread.sortFilterPosts')}
            />
          }
          id="postsFilter-dropdown"
        >
          <PostsFilterLabelMenuItem labelMsgId="debate.thread.sortPosts" />

          {postsOrderPolicies.map(item => (
            <PostsFilterMenuItem
              key={item.id}
              item={item}
              selected={item.id === selectedPostsOrderPolicy.id}
              inputType="radio"
              inputName="postsOrder"
              onSelectItem={this.selectPostsOrderPolicy}
              eventKey={item.id}
            />
          ))}
          <PostsFilterLabelMenuItem labelMsgId="debate.thread.overviewPosts" />

          {postsDisplayPolicies.map(item => (
            <PostsFilterMenuItem
              key={item.id}
              item={item}
              selected={item.id === selectedPostsDisplayPolicy.id}
              inputType="radio"
              inputName="postsDisplay"
              onSelectItem={this.selectPostsDisplayPolicy}
              eventKey={item.id}
            />
          ))}

          <PostsFilterButtons>
            <PostsFilterButton
              id="postsFilter-button-reset"
              onClick={this.resetPolicies}
              i18nTitle="debate.thread.postsOrder.cleanFilter"
              bsStyle="default"
            />
            <PostsFilterButton
              id="postsFilter-button-save"
              style={{ textAlign: 'right', float: 'right' }}
              onClick={this.savePolicies}
              i18nTitle="ok"
              bsStyle="success"
            />
          </PostsFilterButtons>
        </DropdownButton>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  setPostsFilterPolicies: (postsDisplay: PostsDisplayPolicy, postsOrder: PostsOrderPolicy) =>
    dispatch(setThreadPostsFilterPolicies(postsDisplay, postsOrder))
});

const mapStateToProps = (state) => {
  const { postsOrderPolicy, postsDisplayPolicy } = state.threadFilter;
  return {
    postsOrderPolicy: postsOrderPolicy,
    postsDisplayPolicy: postsDisplayPolicy
  };
};

export default withScreenHeight(connect(mapStateToProps, mapDispatchToProps)(DumbPostsFilterMenu));