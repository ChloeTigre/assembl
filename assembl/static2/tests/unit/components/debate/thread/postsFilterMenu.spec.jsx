import { MockedProvider } from 'react-apollo/test-utils';
import React from 'react';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';

import PostsFilterMenu from '../../../../../js/app/components/debate/common/postsFilter/menu';
import { defaultDisplayPolicy, defaultOrderPolicy } from '../../../../../js/app/components/debate/common/postsFilter/policies';

describe('PostsFilterMenu component', () => {
  const mockStore = configureStore();
  const initialState = {
    threadFilter: {
      postsOrderPolicy: defaultOrderPolicy,
      postsDisplayPolicy: defaultDisplayPolicy
    }
  };

  it('should render a filter with sort by recently started threads selected', () => {
    const props = {};
    const store = mockStore(initialState);
    const component = renderer.create(
      <MockedProvider store={store}>
        <PostsFilterMenu {...props} />
      </MockedProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});