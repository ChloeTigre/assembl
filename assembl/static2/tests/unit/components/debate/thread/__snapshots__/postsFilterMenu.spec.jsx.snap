// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsFilterMenu component should render a filter with sort by recently started threads selected 1`] = `
<div
  className="posts-filter-button"
>
  <div
    className="dropdown btn-group"
  >
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle btn btn-default"
      disabled={false}
      id="postsFilter-dropdown"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      type="button"
    >
      <img
        alt="user pic"
        height={24}
        src="/static2/img/icons/black/filter.svg"
        width={24}
      />
       
      <span
        className="caret"
      />
    </button>
    <ul
      aria-labelledby="postsFilter-dropdown"
      className="dropdown-menu dropdown-menu-right"
      role="menu"
    >
      <li
        className="disabled"
        role="presentation"
      >
        <a
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          <strong>
            <span>
              Sort
            </span>
          </strong>
        </a>
      </li>
      <li
        className=""
        role="presentation"
      >
        <a
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          <input
            checked={false}
            id="postsFilterItem-reverseChronologicalTop"
            name="postsOrder"
            readOnly={true}
            type="radio"
          />
           
          <label
            htmlFor="postsFilterItem-reverseChronologicalTop"
          >
            <span>
              Recently active threads
            </span>
          </label>
        </a>
      </li>
      <li
        className=""
        role="presentation"
      >
        <a
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          <input
            checked={true}
            id="postsFilterItem-reverseChronologicalLast"
            name="postsOrder"
            readOnly={true}
            type="radio"
          />
           
          <label
            htmlFor="postsFilterItem-reverseChronologicalLast"
          >
            <span>
              Recently started threads
            </span>
          </label>
        </a>
      </li>
      <li
        className=""
        role="presentation"
      >
        <a
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          <input
            checked={false}
            id="postsFilterItem-chronologicalTop"
            name="postsOrder"
            readOnly={true}
            type="radio"
          />
           
          <label
            htmlFor="postsFilterItem-chronologicalTop"
          >
            <span>
              Chronological threads
            </span>
          </label>
        </a>
      </li>
      <li
        className=""
        role="presentation"
      >
        <a
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          <input
            checked={false}
            id="postsFilterItem-reverseChronologicalFlat"
            name="postsOrder"
            readOnly={true}
            type="radio"
          />
           
          <label
            htmlFor="postsFilterItem-reverseChronologicalFlat"
          >
            <span>
              Newest messages first
            </span>
          </label>
        </a>
      </li>
      <li
        className=""
        role="presentation"
      >
        <a
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          <input
            checked={false}
            id="postsFilterItem-popularityFlat"
            name="postsOrder"
            readOnly={true}
            type="radio"
          />
           
          <label
            htmlFor="postsFilterItem-popularityFlat"
          >
            <span>
              Most popular messages first
            </span>
          </label>
        </a>
      </li>
      <li
        className="disabled"
        role="presentation"
      >
        <a
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          <strong>
            <span>
              Overview
            </span>
          </strong>
        </a>
      </li>
      <li
        className=""
        role="presentation"
      >
        <a
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          <input
            checked={true}
            id="postsFilterItem-full"
            name="postsDisplay"
            readOnly={true}
            type="radio"
          />
           
          <label
            htmlFor="postsFilterItem-full"
          >
            <span>
              Full messages
            </span>
          </label>
        </a>
      </li>
      <li
        className=""
        role="presentation"
      >
        <a
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          <input
            checked={false}
            id="postsFilterItem-summary"
            name="postsDisplay"
            readOnly={true}
            type="radio"
          />
           
          <label
            htmlFor="postsFilterItem-summary"
          >
            <span>
              Message previews
            </span>
          </label>
        </a>
      </li>
      <li
        className=""
        role="presentation"
      >
        <a
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          <button
            className="pull-right btn btn-primary"
            disabled={false}
            onClick={[Function]}
            onKeyDown={[Function]}
            onSelect={[Function]}
            type="button"
          >
            <span>
              Clean filters
            </span>
          </button>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
